### PR TITLE
Feature/parser optimizations

### PR DESCRIPTION
**Pull request by xunilrj**
part of #1837 

This PR do two very fine grained optimizations, it does not solve the "parser forking" problem. But results are sufficiently nice to be merged and push solving the problem for later.

- Parser::events are now using an immutable vector ()
  - actually a RRB trees from https://github.com/bodil/im-rs 
  - This video explains what Radix Balanced Tree are (and its relaxed version): https://youtu.be/sPhpelUfu8Q?t=563 around 9:20
  - https://youtu.be/cUx2b_FO8EQ?t=2679 at 44:30
- TokenSource is split in two and everything that can be shared in behind an Arc.

# How to test

```
> cargo run -p xtask --release -- coverage-libs --filter=three
    Finished release [optimized] target(s) in 1.73s
     Running `target/release/xtask coverage-libs --filter=three`
[three.min.js] - using [target/three.min.js]
Tokenizer
        Time
                took 28.460508ms
Parser
        Time
                took 1.365925465s
TreeSink
        Time
                took 37.974648ms
Total Time: 1.436760875s
end

> cargo run -p xtask --release -- coverage-libs | grep "Total Time:"
    Finished release [optimized] target(s) in 1.52s
     Running `target/release/xtask coverage-libs`
Total Time: 76.490934ms
Total Time: 141.972642ms
Total Time: 1.076629ms
Total Time: 69.732935ms
Total Time: 484.809849ms
Total Time: 7.937441ms
Total Time: 1.033502967s
Total Time: 667.98844ms
Total Time: 968.266103ms
```

After
```
> cargo run -p xtask --release -- coverage-libs --filter=three
    Finished release [optimized] target(s) in 1.62s
     Running `target/release/xtask coverage-libs --filter=three`
[three.min.js] - using [target/three.min.js]
Tokenizer
        Time
                took 26.19684ms
Parser
        Time
                took 62.691038ms
TreeSink
        Time
                took 55.820578ms
Total Time: 158.824674ms
end
> cargo run -p xtask --release -- coverage-libs | grep "Total Time:"
    Finished release [optimized] target(s) in 1.30s
     Running `target/release/xtask coverage-libs`
Total Time: 39.468351ms
Total Time: 34.644993ms
Total Time: 1.184318ms
Total Time: 21.686041ms
Total Time: 128.342229ms
Total Time: 5.652955ms
Total Time: 67.573408ms
Total Time: 73.409934ms
Total Time: 96.870615ms
```

Direct comparison:

```
Total Time: 76.490934ms    -> 39.468351ms
Total Time: 141.972642ms  -> 34.644993ms
Total Time: 1.076629ms      ->  1.184318ms
Total Time: 69.732935ms    -> 21.686041ms
Total Time: 484.809849ms  -> 128.342229ms
Total Time: 7.937441ms      -> 5.652955ms
Total Time: 1.033502967s   -> 67.573408ms
Total Time: 667.98844ms    -> 73.409934ms
Total Time: 968.266103ms  -> 96.870615ms
```

It is also interesting to see the difference in memory consumption:

Before
```
[three.min.js] - using [target/three.min.js]
Tokenizer
        Memory
                Current Blocks: 51
                Current Bytes: 9.26 MB
                Max Blocks: 51
                Max Bytes: 9.26 MB
                Total Blocks: 24
                Total Bytes: 14.99 MB
Parser
        Memory
                Current Blocks: 86
                Current Bytes: 32.59 MB
                Max Blocks: 143
                Max Bytes: 55.60 MB
                Total Blocks: 182398
                Total Bytes: 15.98 GB
TreeSink
        Memory
                Current Blocks: 84192
                Current Bytes: 10.31 MB
                Max Blocks: 143
                Max Bytes: 55.60 MB
                Total Blocks: 91070
                Total Bytes: 7.51 MB
```

After
```
[three.min.js] - using [target/three.min.js]
Tokenizer
        Memory
                Current Blocks: 52
                Current Bytes: 9.26 MB
                Max Blocks: 52
                Max Bytes: 9.26 MB
                Total Blocks: 25
                Total Bytes: 14.99 MB
Parser
        Memory
                Current Blocks: 9877
                Current Bytes: 19.31 MB
                Max Blocks: 9887
                Max Bytes: 24.02 MB
                Total Blocks: 194405
                Total Bytes: 550.22 MB
TreeSink
        Time
                took 769.558321ms
        Memory
                Current Blocks: 84192
                Current Bytes: 10.31 MB
                Max Blocks: 93987
                Max Bytes: 25.71 MB
                Total Blocks: 91071
                Total Bytes: 7.51 MB
```

Interesting values are:
```
Final: 10.31 MB -> 10.31 MB
Max: 55.60 MB -> 25.71 MB
Max Total: 15.98 GB -> 550.22 MB
```

### CODE REVIEW

**Comment by IWANABETHATGUY** [comment]
```
use std::time::Instant;

use rslint_parser::{
    ast::JsRoot, process, syntax::program, tokenize, AstNode, LosslessTreeSink,
    Parser, Syntax, SyntaxNode, TokenSource,
};
fn main() {
    let source = include_str!("../test.js");

    let start = Instant::now();    // File id is used for the labels inside parser errors to report them, the file id
    // is used to look up a file's source code and path inside of a codespan `Files` implementation.
    let (tokens, lexer_errors) = tokenize(source, 0);

    assert!(lexer_errors.is_empty());

    // The parser uses a token source which manages yielding non-whitespace tokens,
    // and giving raw tokens to allow the parser to turn completed markers into syntax nodes
    let token_source = TokenSource::new(source, &tokens);

    let mut parser = Parser::new(token_source, 0, Syntax::new(rslint_parser::FileKind::Module));

    // Use one of the syntax parsing functions to parse an expression.
    // This adds node and token events to the parser which are then used to make a node.
    // A completed marker marks the start and end indices in the events vec which signify
    // the Start event, and the Finish event.
    // Completed markers can be turned into an ast node with parse_marker on the parser
    let completed_marker = program::parse(&mut parser);

    parser.parse_marker::<JsRoot>(&completed_marker);

    // Make a new text tree sink, its job is assembling events into a rowan GreenNode.
    // At each point (Start, Token, Finish, Error) it also consumes whitespace.
    // Other abstractions can also yield lossy syntax nodes if whitespace is not wanted.
    // Swap this for a LossyTreeSink for a lossy AST result.
    let mut sink = LosslessTreeSink::new(source, &tokens);

    // Consume the parser and get its events, then apply them to a tree sink with `process`.
    let (events, errors) = parser.finish();
    process(&mut sink, events, errors);

    let (untyped_node, errors) = sink.finish();

    println!("{:?}", errors);

    assert!(JsRoot::can_cast(untyped_node.kind()));

    // Convert the untyped SyntaxNode into a typed AST node
    let _typed_expr = JsRoot::cast(untyped_node).unwrap();
    println!("{:?}", start.elapsed());
    // println!("{}", typed_expr);
}

```
use the same code snippet, this branch is slower than the main branch, i don't know why,
main branch is about 6ms on my computer,
this branch took about 8-9ms

**Comment by IWANABETHATGUY** [comment]
MacBook Pro (16-inch, 2019)
2.6 GHz 六核Intel Core i7

**Comment by xunilrj** [comment]
> use the same code snippet, this branch is slower than the main branch, i don't know why, main branch is about 6ms on my computer, this branch took about 8-9ms

It is consistently slower? Because this difference is so small that can be noise. I see bigger variance when parsing some libs, for example. If you look at the results above, some smaller libs (React) is actually a little bit slower in this branch.

The TokenSource in this branch should be always faster; but pushing to Parser::events is now slower than before (only cloning is much faster).

File: crates/rslint_parser/src/parser.rs [hunk 1]
```diff
@@ -79,7 +79,7 @@ use crate::*;
 pub struct Parser<'t> {
 	pub file_id: usize,
 	tokens: TokenSource<'t>,
-	pub(crate) events: Vec<Event>,
+	pub(crate) events: im_rc::Vector<Event>,
```
[the remark this answers is not in the document]
Comment (xunilrj) [reply crates/rslint_parser/src/parser.rs:122]: Maybe we could, but ```Rc``` alone is not enough, I would need to use ``Rc<RefCel<...>>```, and ```RefCell`` has runtime implications on every access. Probably still faster than the RRB tree, but I would also need to change every access to Parser::events.

File: crates/rslint_parser/src/token_source.rs [hunk 2]
```diff
@@ -14,6 +14,12 @@ pub struct TokenSource<'t> {
 	token_offset_pairs: Vec<(rslint_lexer::Token, TextSize)>,
 	/// A list of the tokens including whitespace.
 	pub raw_tokens: &'t [rslint_lexer::Token],
+}
+
+/// The source of tokens for the parser
+#[derive(Clone)]
+pub struct TokenSource<'t> {
+	pub data: Arc<TokenSourceData<'t>>,
```
[the remark this answers is not in the document]
Comment (xunilrj) [reply crates/rslint_parser/src/token_source.rs]: Done.

**Comment by IWANABETHATGUY** [comment]
> > use the same code snippet, this branch is slower than the main branch, i don't know why, main branch is about 6ms on my computer, this branch took about 8-9ms
> 
> It is consistently slower? Because this difference is so small that can be noise. I see bigger variance when parsing some libs, for example. If you look at the results above, some smaller libs (React) is actually a little bit slower in this branch.
> 
> The TokenSource in this branch should be always faster; but pushing to Parser::events is now slower than before (only cloning is much faster).

yes, my test.js is about 900 loc

**Pull request closed without merging**